File: LeetCode/2043/python/lc2043.py
from typing import List

# Subjective level: easy
class Bank:
    # Runtime complexity: O(n)
    # Auxiliary space complexity: O(n)
    def __init__(self, balance: List[int]):
        self.accounts = dict()
        for i in range(len(balance)):
            self.accounts[i+1] = balance[i]
        print(self.accounts)

    # Runtime complexity: O(1)
    # Auxiliary space complexity: O(1)
    def transfer(self, fromAccount: int, toAccount: int, amount: int):
        if fromAccount in self.accounts.keys() and toAccount in self.accounts.keys():
            if self.accounts[fromAccount] >= amount:
                self.accounts[fromAccount] -= amount
                self.accounts[toAccount] += amount
                return True
        return False

    # Runtime complexity: O(1)
    # Auxiliary space complexity: O(1)
    def deposit(self, account: int, amount: int):
        if account in self.accounts.keys():
            self.accounts[account] += amount
            return True
        return False

    # Runtime complexity: O(1)
    # Auxiliary space complexity: O(1)
    def withdraw(self, account: int, amount: int):
        if account in self.accounts.keys():
            if self.accounts[account] >= amount:
                self.accounts[account] -= amount
                return True
        return False

File: LeetCode/2043/golang/lc2043.go
package lc2043

// Subjective level: easy+
// Solved on: 2026-01-25

type Bank struct {
	Accounts map[int]int64
}

func Constructor(balances []int64) Bank {
	ans := Bank{}
	ans.Accounts = make(map[int]int64)
	for i, balance := range balances {
		// Note: balances is 0-indexed, but initially bank uses 1-indexed account numbers,
		// so if there are n entries, these have to be numbered from 1 to n.
		ans.Accounts[i+1] = balance
	}
	return ans
}

// Runtime complexity: O(1)
// Auxiliary space: O(1)
func (this *Bank) Transfer(fromAccount int, toAccount int, amount int64) bool {
	_, fromAccountExists := this.Accounts[fromAccount]
	_, toAccountExists := this.Accounts[toAccount]
	if !(fromAccountExists && toAccountExists) {
		return false
	}
	if this.Accounts[fromAccount] < amount {
		return false
	}

	this.Accounts[fromAccount] -= amount
	this.Accounts[toAccount] += amount
	return true
}

// Runtime complexity: O(1)
// Auxiliary space: O(1)
func (this *Bank) Deposit(account int, amount int64) bool {
	_, accountExists := this.Accounts[account]
	if !accountExists {
		return false
	}
	this.Accounts[account] += amount
	return true
}

// Runtime complexity: O(1)
// Auxiliary space: O(1)
func (this *Bank) Withdraw(account int, amount int64) bool {
	_, accountExists := this.Accounts[account]
	if !accountExists {
		return false
	}
	if this.Accounts[account] < amount {
		return false
	}
	this.Accounts[account] -= amount
	return true
}
